test/integration/job/job_test.go

```diff
@@ -1805,41 +1823,25 @@ func TestJobPodReplacementPolicy(t *testing.T) {
 			}
 			jobClient := clientSet.BatchV1().Jobs(jobObj.Namespace)
 
-			// Wait for pods to start up.
-			err = wait.PollUntilContextTimeout(ctx, 5*time.Millisecond, wait.ForeverTestTimeout, true, func(ctx context.Context) (done bool, err error) {
-				job, err := jobClient.Get(ctx, jobObj.Name, metav1.GetOptions{})
-				if err != nil {
-					return false, err
-				}
-				if job.Status.Active == podCount {
-					return true, nil
-				}
-				return false, nil
+			waitForPodsToBeActive(ctx, t, jobClient, 2, jobObj)
+
+			addFinalizerAndDeletePods(ctx, t, clientSet, ns.Name)
+
+			validateJobsPodsStatusOnly(ctx, t, clientSet, jobObj, podsByStatus{
+				Terminating: tc.wantStatusAfterDeletion.terminating,
+				Failed:      tc.wantStatusAfterDeletion.failed,
+				Active:      tc.wantStatusAfterDeletion.active,
+				Ready:       pointer.Int32(0),
 			})
-			if err != nil {
-				t.Fatalf("Error waiting for Job pods to become active: %v", err)
-			}
-			if tc.deletePods {
-				err = clientSet.CoreV1().Pods(ns.Name).DeleteCollection(ctx,
-					metav1.DeleteOptions{},
-					metav1.ListOptions{
-						Limit: 1000,
-					})
-				if err != nil {
-					t.Fatalf("Failed to delete Pods: %v", err)
-				}
-			}
-			if tc.failPods {
-				err, _ = setJobPodsPhase(ctx, clientSet, jobObj, v1.PodFailed, int(podCount))
-				if err != nil {
-					t.Fatalf("Failed setting phase %s on Job Pods: %v", v1.PodFailed, err)
-				}
-			}
+
+			failTerminatingPods(ctx, t, clientSet, ns.Name)
+
+			removePodsFinalizer(ctx, t, clientSet, ns.Name)
```

Comment:
```suggestion
			t.Cleanup(removePodsFinalizer(ctx, t, clientSet, ns.Name))
```

Although, it should be called right after we know that the pods are active (so that the finalizer is removed even if the test fails)

Reply:
do you mean still call it at line 1839 but also call it in `t.Cleanup` after pods are active?

Reply:
hmmm calling the `removePodsFinalizer` results in an error when listing pods:
```
client rate limiter Wait returned an error: context canceled
```

Reply:
that makes sense, because we close the context.

use context.Background just for the cleanup.

Reply:
just to clarify, should I have 2 calls for `removePodFinalizer`, one in the current place and one in `t.Cleanup`?

Reply:
I think the controller shuts down before the `t.Cleanup`. The error I get if I use a background context is:
```
    job_test.go:1878: Failed to list pods: Get "https://127.0.0.1:56263/api/v1/namespaces/pod-replacement-policy/pods": dial tcp 127.0.0.1:56263: connect: connection refused
```

Reply:
The problem is that we are mixing `t.Cleanup` and `defer` (the defers would run before the cleanups). If you move everything to `t.Cleanup`, it works.
```gitdiff
diff --git a/test/integration/job/job_test.go b/test/integration/job/job_test.go
index 5921abd7aac..1d6ab72ac44 100644
--- a/test/integration/job/job_test.go
+++ b/test/integration/job/job_test.go
@@ -1861,9 +1861,9 @@ func TestJobPodReplacementPolicy(t *testing.T) {
                        defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobPodFailurePolicy, tc.jobSpec.PodFailurePolicy != nil)()
 
                        closeFn, restConfig, clientSet, ns := setup(t, "pod-replacement-policy")
-                       defer closeFn()
+                       t.Cleanup(closeFn)
                        ctx, cancel := startJobControllerAndWaitForCaches(t, restConfig)
-                       defer cancel()
+                       t.Cleanup(cancel)
                        resetMetrics()
 
                        jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &batchv1.Job{
@@ -1875,6 +1875,9 @@ func TestJobPodReplacementPolicy(t *testing.T) {
                        jobClient := clientSet.BatchV1().Jobs(jobObj.Namespace)
 
                        waitForPodsToBeActive(ctx, t, jobClient, 2, jobObj)
+                       t.Cleanup(func() {
+                               removePodsFinalizer(ctx, t, clientSet, ns.Name)
+                       })
 
                        deletePods(ctx, t, clientSet, ns.Name)
 
@@ -1887,8 +1890,6 @@ func TestJobPodReplacementPolicy(t *testing.T) {
 
                        failTerminatingPods(ctx, t, clientSet, ns.Name)
 
-                       removePodsFinalizer(ctx, t, clientSet, ns.Name)
-
                        validateJobsPodsStatusOnly(ctx, t, clientSet, jobObj, podsByStatus{
                                Terminating: tc.wantStatusAfterFailure.terminating,
                                Failed:      tc.wantStatusAfterFailure.failed,
```

Reply:
makes sense @alculquicondor, it works now, thanks for the suggestion